src/Function.cpp

```diff
@@ -77,6 +77,59 @@ static long cur_func_idx;				// Index into FuncList that we are currently workin
 static bool param_first=true;			// Flag to track output of commas
 static int builtin_functions_cnt;
 
+FunctionAttribute::FunctionAttribute(string name, int prob)
+	:  attribute(name), attribute_probability(prob)
+{
+	//Nothing else to do
+}
+
+void
+FunctionAttribute::OutputAttribute(std::ostream &out, string option)
+{
+	out << " __attribute__((" << attribute << option << "))";
+}
+
+BooleanFunctionAttribute::BooleanFunctionAttribute(string name, int prob)
+	: FunctionAttribute(name, prob)
+{
+	//Nothing else to do
+}
+
+void
+BooleanFunctionAttribute::OutputAttributes(std::ostream &out)
+{
+	if(rnd_flipcoin(attribute_probability))
+		OutputAttribute(out, "");
+}
+
+MultiValuedFunctionAttribute::MultiValuedFunctionAttribute(string name, int prob, vector<string> arguments)
+	: FunctionAttribute(name, prob)
+{
+	attribute_values = arguments;
+}
+
+void
+MultiValuedFunctionAttribute::OutputAttributes(std::ostream &out)
+{
+	if(rnd_flipcoin(attribute_probability))
+		OutputAttribute(out, attribute_values[rnd_upto(attribute_values.size())]);
+}
+
+void
+Function::GenerateAttributes(Function *f)
+{
+	if(CGOptions::func_attr_flag()){
+		common_func_attributes = {"artificial", "flatten", "no_reorder", "hot", "cold", "no_ipa", "no_throw", "deprecated", "no_icf", "no_profile_instrument_function", "no_instrument_function", "no_sanitize_address", "no_sanitize_thread", "no_sanitize_undefined", "no_split_stack", "noinline", "noplt", "stack_protect"};
+		vector<string>::iterator itr;
+		for(itr = f->common_func_attributes.begin(); itr < f->common_func_attributes.end(); itr++)
+			f->attributes.push_back(new BooleanFunctionAttribute(*itr, FuncAttrProb));
+
+		f->attributes.push_back(new MultiValuedFunctionAttribute("visibility", FuncAttrProb, {"(\"default\")", "(\"hidden\")", "(\"protected\")", "(\"internal\")"}));
```

Comment:
Please do not wrap the string with `("` and `")`, but always wrap the options in `OutputAttributes`.

Reply:
In `OutputAttributes`, I am wrapping it like -
`"(\"" +attribute_values[rnd_upto(attribute_values.size())] + "\")"` as second parameter to `OutputAttribute`
Is this fine?

Reply:
Please do not ask me about each small step. Make it best you can and I'll provide then a review.

Reply:
Sure. I have done all the suggested changes required. Should I push changes now?

Reply:
Yes please.